Writer

ibrechin opened the pull request:
Added the capability to write BAI2 files. Not in a necessarily particularly clever way but it more or less works.

ibrechin committed:
Add capability for writing BAI2 files to library

These additions allow one to write out a valid BAI2 file,
after having constructed valid model objects.

The aim is that it should at least support writing every
format of BAI2 file that it can parse.

ibrechin committed:
Automatically set trailer totals when writing file

ibrechin committed:
Set defaults on model creation

ibrechin committed:
Change availability from tuple-list to OrderedDict

ibrechin committed:
Fix writing distributed availability funds

ibrechin committed:
Add tests for file writing functionality and fix issues

Support for the different Funds Type values in the
Account Identifier record.

ibrechin committed:
Add support for clock format time

The NatWest BAI2 files include times in the format HH:MM:SS,
although they still used '2400' as the end of day value.

ibrechin committed:
Update model creation and writing interface

ibrechin committed:
Update README with instructions for writing BAI2 files

ibrechin commented:
@ushkarev @SteveMarshall have a wee look

ibrechin committed:
Ensure that account control total includes summary amounts

The Account control total in the account trailer record
should also include all of the amounts contained in the
summary sections of the account identifier record.

SteveMarshall commented on tests/test_bai2.py:
```diff
@@ -56,6 +56,15 @@ def test_parse_from_file(self):
+        with open(file_path) as f:
+            bai2_file = bai2.parse_from_file(f)
+            self.assertTrue(isinstance(bai2_file, Bai2File))
+
```
What's the difference between `example.bai2` and `example2.bai2`, in terms of the data they contain? Is there something more we can say about what they are/where they're from?

ibrechin replied:
Have renamed to show which bank each is from.

SteveMarshall commented on tests/test_utils.py:
```diff
@@ -48,3 +68,39 @@ def test_parse_beginning_of_day(self):
+        self.assertEqual(str_value, '1759')
+
+    def test_write_intra_day_time_with_setting_true(self, mock_environ):
+        mock_environ.USE_CLOCK_FORMAT_FOR_INTRA_DAY.return_value = True
```
Why `.return_value` here and `=` in the other two?

SteveMarshall commented on tests/test_utils.py:
```diff
@@ -48,3 +68,39 @@ def test_parse_beginning_of_day(self):
+
+
+@mock.patch('bai2.utils.settings')
+class WriteTime(TestCase):
```
Is there ever a time that we should write a `9999`?

ibrechin replied:
Unlikely. If there is, we'll handle it then.

SteveMarshall commented on tests/test_writers.py:
```diff
@@ -0,0 +1,406 @@
+            [],
+            type_code=constants.TypeCodes['399'],
+            amount=2599,
+            text='BILLS BILLS BILLS BILLS BILLS BILLS BILLS BILLS BILLS BILLS BILLS BILLS',
```
:dollar: :dollar: bill y'all.

SteveMarshall commented on tests/test_writers.py:
```diff
@@ -0,0 +1,406 @@
+from unittest import TestCase
```
Might be nice to break this up into `tests/transaction_writers/detail.py` and so on.

ibrechin replied:
I think that's excessive nesting.

SteveMarshall commented:
I'd like the tests throughout this to have more explicit names about what they're testing. It would be easier for others to understand the intent of the tests if the names gave a better impression of what the expected outcome was: as it is, they feel a bit like half-finished sentences.

For example, `TransactionDetailWriterTestCase.test_transaction_detail_with_default_availability` might be better named as`TransactionDetailWriterTestCase.test_default_availability_renders_correctly`.

SteveMarshall commented on README.rst:
```diff
@@ -79,6 +97,9 @@ Settings
  * **BAI2_IGNORE_INTEGRITY_CHECKS**: if `1`, it disables integrity checks.
+ * **BAI2_TEXT_ON_NEW_LINE**: if `1`, the TEXT field of a transaction always begins a new continuation record.
+ * **BAI2_LINE_LENGTH**: the maximum length of a transaction or account identifier record before a new continuation record will be created.
+ * **BAI2_USE_CLOCK_FORMAT_FOR_INTRA_DAY**: if `1`, uses the format HH:MM:SS instead of HHMM for times which are not at the end of the day (which will still be 2400).
```
I wonder if these might be better as parameters passed in on construction or at write time? It seems a little odd to tie parser/writer behaviours to the environment… (I might say as much of the integrity checks, in hindsight…)

ibrechin replied:
That is also a thought that I had. I can't remember why I didn't do it. I'll do it.

ibrechin committed:
Make the names of tests and test data clearer

SteveMarshall commented on bai2/writers.py:
```diff
@@ -0,0 +1,274 @@
+
+        self.obj.update_totals()
+        self.obj.trailer.number_of_records = len(
+            '\n'.join([header] + children).split('\n')) + 1
```
This feels a bit… convoluted to me, joining and then re-splitting. What's the difference between doing this and using `len(children)`? Or are the children going to have `\n`s in them?

ibrechin replied:
Yeah, it's not great.

It's because of continuation records. Each record writer writes itself out as a single complete string, including any continuation records, whereas the record count at the end needs to count continuation records separately.

Making each writer return a list of lines instead of just a string is something that I considered doing but didn't, because of reasons. I can't remember what they were. I might try again.

ibrechin committed:
Move configuration from env variables to function args

ibrechin committed:
Refactor writers to output lists of records

This is instead of outputting a string containing newlines. This
makes it cleaner to reason about the structure of the file whilst
writing it.

SteveMarshall commented:
Looks good to me! @ushkarev?
